index-scheduler/src/lib.rs

```diff
@@ -561,11 +563,73 @@ impl IndexScheduler {
         Ok(tasks)
     }
 
-    /// Returns the tasks matched by the given query.
-    pub fn get_tasks(&self, query: Query) -> Result<Vec<Task>> {
-        let tasks = self.get_task_ids(&query)?;
+    /// Return true iff there is at least one task associated with this index
+    /// that is processing.
+    pub fn is_index_processing(&self, index: &str) -> Result<bool> {
+        let rtxn = self.env.read_txn()?;
+        let processing_tasks = self.processing_tasks.read().unwrap().processing.clone();
+        let index_tasks = self.index_tasks(&rtxn, index)?;
+
+        let nbr_index_processing_tasks = processing_tasks.intersection_len(&index_tasks);
+
+        Ok(nbr_index_processing_tasks > 0)
+    }
+
+    /// Return the task ids matching the query from the user's point of view.
+    ///
+    /// There are two differences between an internal query and a query executed by
+    /// the user.
+    ///
+    /// 1. IndexSwap tasks are not publicly associated with any index, but they are associated
+    /// with many indexes internally.
+    /// 2. The user may not have the rights to access the tasks (internally) associated wuth all indexes.
+    pub fn get_task_ids_from_authorized_indexes(
+        &self,
+        rtxn: &RoTxn,
+        query: &Query,
+        authorized_indexes: &Option<Vec<String>>,
+    ) -> Result<RoaringBitmap> {
+        let mut tasks = self.get_task_ids(rtxn, query)?;
+
+        // If the query contains a list of index_uid, then we must exclude IndexSwap tasks
+        // from the result (because it is not publicly associated with any index)
+        if query.index_uid.is_some() {
+            tasks -= self.get_kind(rtxn, Kind::IndexSwap)?
+        }
```

Comment:
Ok, so that’s pretty good. But actually, after testing the old meilisearch a little bit, I found out we’re supposed to ignore *every* kind of task that is not associated with an index_uid.
Thus:
- TaskCancelation
- TaskDeletion
- DumpCreation (only this one was present in the previous version of meilisearch and I missed it)
- SnapshotCreation

Reply:
Ok so ACTUALLY after rerethinking about it with @gmourier it’s a little bit more complicated.

When checking the allowed indexes **by the API-keys**, if there is a limitation on the indexes (`index_uid!=*` or `None` in the code I think).
We must filter out the « global tasks »:
- TaskCancelation
- TaskDeletion
- DumpCreation

In the case of the `IndexSwap` we can see it only if in the user query we typed `index_uid=*` so this is good.

Now the bad thing I think is... what if there is a swap(A, B) and your API key only allows you to access to A?
You should not see the swap but here from what I understand of your code we will receive it :thinking:
And I don't really knows how to do this efficiently...

Maybe @Kerollmops you'll have an idea?

Reply:
I am not sure I understand the problem.
If someone specifies `indexUid=catto`, then we'll hit line 495:
```rust
        if let Some(index) = &query.index_uid {
            let mut index_tasks = RoaringBitmap::new();
            for index in index {
                index_tasks |= self.index_tasks(rtxn, index)?;
            }
            tasks &= &index_tasks;
        }
```
and the taskDeletion, taskCancelation, etc. won't be returned.

If there is a `indexSwap(A, B)` and the API key only allows me to access A, then we'll hit line 605:
```rust
      // Any task that is internally associated with a non-authorized index
        // must be discarded.
        if let Some(authorized_indexes) = authorized_indexes {
            let all_indexes_iter = self.index_tasks.iter(rtxn)?;
            for result in all_indexes_iter {
                let (index, index_tasks) = result?;
                if !authorized_indexes.contains(&index.to_owned()) {
                    tasks -= index_tasks;
                }
            }
        }
```
which will remove the task from the results as well. (In the loop there, `index == B` and `authorised_indexes == [A]`, so `[A].contains(B)` will return false and we'll run `tasks -= [indexSwap(A,B)]`)